Patient Timeline Improvements

Luke-Sikina opened the pull request:
Fix cBioPortal/cbioportal#6882

Describe changes proposed in this pull request:

- Add TSV download
- Update version of clinical-timeline to 0.0.30. This comes with the following enhancements:
  - Cut off circles
    - Issue: mousing over points on the bottom lane of the
    timeline caused them to be cut off when they expanded.
    - Fix: added 3 pixels to the height of the svg

Test with

https://deploy-preview-2950--cbioportalfrontend.netlify.com/patient?caseId=P29&studyId=lgg_ucsf_2014#navCaseIds=lgg_ucsf_2014:P01,lgg_ucsf_2014:P02,lgg_ucsf_2014:P04,lgg_ucsf_2014:P05,lgg_ucsf_2014:P06,lgg_ucsf_2014:P07,lgg_ucsf_2014:P08,lgg_ucsf_2014:P09,lgg_ucsf_2014:P10,lgg_ucsf_2014:P11,lgg_ucsf_2014:P12,lgg_ucsf_2014:P13,lgg_ucsf_2014:P15,lgg_ucsf_2014:P16,lgg_ucsf_2014:P17,lgg_ucsf_2014:P18,lgg_ucsf_2014:P21,lgg_ucsf_2014:P24,lgg_ucsf_2014:P25,lgg_ucsf_2014:P26,lgg_ucsf_2014:P27,lgg_ucsf_2014:P28,lgg_ucsf_2014:P29

Also possible to test on private portals:

```
localStorage.clear()
localStorage.netlify = 'deploy-preview-2950--cbioportalfrontend'
```

And refresh the page

Luke-Sikina commented on package.json:
```diff
@@ -118,7 +118,7 @@
     "chart.js": "^2.6.0",
     "classnames": "^2.2.5",
-    "clinical-timeline": "0.0.19",
+    "clinical-timeline": "0.0.20",
```
Presumably this will get updated. The bulk of the work is in that repo, so we need to wait for that to land before this PR can be reviewed and merged.

Link to PR: https://github.com/cBioPortal/clinical-timeline/pull/118

Luke-Sikina commented on src/pages/patientView/timeline/Timeline.tsx:
```diff
@@ -83,7 +97,8 @@ export default class Timeline extends React.Component<ITimelineProps, {}> {
             }
         });
-
+        console.log("building timeline");
```
Delete

Luke-Sikina commented on src/pages/patientView/timeline/legacy.js:
```diff
@@ -61,6 +61,8 @@ export function buildTimeline(params, caseIds, patientInfo, clinicalDataMap, cas
     if (data.length === 0) return;
 
     var timeData = clinicalTimelineExports.clinicalTimelineParser(data);
+    console.log("Data: " + JSON.stringify(timeData));
```
delete

inodb commented:
Nice work!

Not sure if this should be part of the work but i noticed some issues with how the ticks behave at different levels of zoom. They don't seem to update, e.g. if I zoom in to a really tiny part I can never get the ticks to show up every 1 day for instance (smallest tick size seems to be 10 days?).

It works for the default example here: https://rawgit.com/cBioPortal/clinical-timeline/master/index.html

But not for the other example JSON we exchanged before (see slack convo):

![Screen Shot 2020-01-31 at 1.51.55 PM.png](https://images.zenhubusercontent.com/58337ef1b9dc3c622022519c/da2b5a6f-c61d-46ac-a623-d00cb14d2b49) 

Ideally the ticks would be readjusted 👆

It might be that the JSON example triggers a particular corner case

jjgao commented:
How can I test this one?

inodb commented:
Hey @Luke-Sikina - is this the PR we should be testing? The deploy preview on Netlify doesn't seem to be working

inodb commented:
Hi @Luke-Sikina - somehow the zoom doesn't seem to be working for me?

https://deploy-preview-2950--cbioportalfrontend.netlify.com/patient/summary?caseId=P04&studyId=lgg_ucsf_2014

The TSV download seems to be working well 👍

inodb commented:
@Luke-Sikina Noticed some missing parts in the TSV:

Might be good to follow the input spec for the TSV i.e. the data that goes into cBioPortal can be the same format as if u download it from the timeline:

https://docs.cbioportal.org/5.1-data-loading/data-loading/file-formats#timeline-data

I see for instance for P04 that the field Medical Therapy is missing:

https://deploy-preview-2950--cbioportalfrontend.netlify.com/patient/summary?caseId=P04&studyId=lgg_ucsf_2014

The raw input data for that study is here:

https://github.com/cBioPortal/datahub/tree/master/public/lgg_ucsf_2014

I'm not sure if there's a reason for having each track in a separate TSVs there. For a single patient that's hopefully not necessary

inodb commented:
Thanks a ton @Luke-Sikina! Looks great!

Some things I found:

-  [ ] the ticks look good now but it might be nice to use .y.m.d formatting of the ticks when they get > 1 month. E.g. here you see 1200+ days, which would be more comprehensible as something like 1y3m15d. Not sure how tricky this is. There should be a function that in the clinical timeline that does that calculation

    ![Screen Shot 2020-02-13 at 3 52 33 PM](https://user-images.githubusercontent.com/1334004/74477279-fe6fc000-4e78-11ea-9a2e-051804e52423.png)

- [ ] would be nice if the brush had a different color than black  e.g. gray or blueish. Similar to the study view

![Screen Shot 2020-02-13 at 3 57 30 PM](https://user-images.githubusercontent.com/1334004/74477699-c157fd80-4e79-11ea-949d-f4e7bfc8a37a.png)

- [ ] the font for scroll/drag is a little off somehow (like it's not that sharp/crisp):
 
    ![Screen Shot 2020-02-13 at 3 59 45 PM](https://user-images.githubusercontent.com/1334004/74477757-e51b4380-4e79-11ea-83f1-f55fcdef7321.png)

- [ ] there's this kind of strange effect because the scroll/drag text is not part of the mouseover event:

    ![timeline_effect_high_fps](https://user-images.githubusercontent.com/1334004/74477985-6d99e400-4e7a-11ea-923e-60e17e8d12f7.gif)

- [ ] Found some issue with how the beginning+ending is calculated. https://deploy-preview-2950--cbioportalfrontend.netlify.com/patient/summary?caseId=P04&studyId=lgg_ucsf_2014

    ![timeline_calculation_off](https://user-images.githubusercontent.com/1334004/74478213-e305b480-4e7a-11ea-8b46-a73c5afbc74f.gif)

I'm not too worried about most of these. We can file them as a new issue. The first two and the last issue would prolly be good to fix before release. If it turns out some of those are really tricky to fix we can maybe do some workaround. Like enable zoom on demand and indicate that it's in beta or something

Luke-Sikina commented:
@inodb I think I've fixed all the issues. The PR for the timeline repo is here: https://github.com/cBioPortal/clinical-timeline/pull/122

One small note: using patient P04 to test trimming wont work well. There was an issue where the initial timeline wasn't getting trimmed, but when you zoomed in and out the timeline would trim. I fixed that, but P04 was an example where it was happening, so now it doesn't. `P18` from the same study should be a better candidate.

tmazor commented:
@Luke-Sikina There are a few issues with download of the timeline images:
1) If I have zoomed in, the "zoom out" text is visible in the download. Can it be removed?
2) The zig-zag (when present) doesn't download nicely
![image](https://user-images.githubusercontent.com/10508276/75399605-ae5a1a00-58ca-11ea-8dea-22779d5ec928.png)
3) The time labels overlap each other a little bit in the website, and a lot when downloaded
![image](https://user-images.githubusercontent.com/10508276/75399667-db0e3180-58ca-11ea-93d8-77831b443892.png)

The above were tested with P17 (the last one after expanding the zig-zag)
https://deploy-preview-2950--cbioportalfrontend.netlify.com/patient/summary?caseId=P17&studyId=lgg_ucsf_2014

Luke-Sikina commented:
@tmazor 
This is ready for you to test again: https://deploy-preview-2950--cbioportalfrontend.netlify.com/patient?studyId=lgg_ucsf_2014&caseId=P17

The zooming logic has been improved, nothing breaks when you zoom on the edge, and single click zooming works as you expected. 
I couldn't fix the pdf issues- those were already broken, so we'll look to fix them in another pull request

tmazor commented:
@Luke-Sikina The single click zoom is great! 

A few things:
- In the link you provided above, I clicked to zoom on the right side of the plot (so the zig-zag timeline compression was still visible), and then clicked on the zig-zag to expand, at which point the "zoom out" button disappears.
- It appears to only be possible to zoom in once. Can we enable a user to zoom in and then zoom in some more?
- For this case: https://deploy-preview-2950--cbioportalfrontend.netlify.com/patient?studyId=lgg_ucsf_2014&caseId=P18
If I click-and-drag to zoom in from 0-1y, what appears is only 0-~6months

inodb commented:
Thanks so much for reviewing @tmazor !

We are hoping to release this soon. I thought it might be ok to release this without your last two comments fixed and follow up on those later. I've filed the last two things as separate issues: https://github.com/cBioPortal/cbioportal/issues/7325 https://github.com/cBioPortal/cbioportal/issues/7326

Let me know if you think that makes sense and if there's anything else that is required to be fixed before rolling out. Thank you!

inodb commented:
Decided to disable zoom for now because we still found issues unfortunately. We will merge the other parts for now
